src/core/geometry/qgspolyhedralsurface.cpp

```diff
@@ -0,0 +1,997 @@
+/***************************************************************************
+                         qgspolyhedralsurface.cpp
+                         ---------------------
+    begin                : August 2024
+    copyright            : (C) 2024 by Jean Felder
+    email                : jean dot felder at oslandia dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgspolyhedralsurface.h"
+#include "qgsapplication.h"
+#include "qgscurve.h"
+#include "qgsfeedback.h"
+#include "qgsgeometryutils.h"
+#include "qgslinestring.h"
+#include "qgslogger.h"
+#include "qgsmultipolygon.h"
+#include "qgspolygon.h"
+#include "qgsvertexid.h"
+#include "qgswkbptr.h"
+#include "qgsmultilinestring.h"
+
+#include <QPainter>
+#include <QPainterPath>
+#include <memory>
+#include <nlohmann/json.hpp>
+
+QgsPolyhedralSurface::QgsPolyhedralSurface()
+{
+  mWkbType = Qgis::WkbType::PolyhedralSurface;
+}
+
+QgsPolyhedralSurface::QgsPolyhedralSurface( QgsMultiPolygon *multiPolygon )
+{
+  if ( multiPolygon->isEmpty() )
+  {
+    return;
+  }
+
+  mPatches.reserve( multiPolygon->numGeometries() );
+  for ( int i = 0; i < multiPolygon->numGeometries(); ++i )
+  {
+    mPatches.append( multiPolygon->polygonN( i ) );
```

Comment:
This is certainly going to crash -- you've marked this method as transferring ownership of the multiPolygon, but at the end of this method the multiPolygon still owns all the parts returned by polygonN... which are now also owned by this object.

I would instead remove the ownership transferral from this method, make the multiPolygon argument const, and use a clone here instead.

Reply:
Fixed.